backend/src/categories/categories.controller.ts

```diff
@@ -1,25 +1,142 @@
+import { In } from 'typeorm';
 @Controller()
 export class CategoriesController {
-  constructor(private categoriesService: CategoriesService) {}
-
+  constructor(
+    private categoriesService: CategoriesService,
+    private wordsService: WordsService,
+    private choicesService: ChoicesService,
+  ) {}
+  // Reminder: Add authguard
   @Get(['student/categories', 'admin/categories'])
   async all() {
     return this.categoriesService.find({});
   }
-
-  @Get(['student/categories/words', 'admin/categories/words'])
-  async allWords() {
-    return this.categoriesService.find({
+  // Reminder: Add authguard
+  @Get(['student/:categoryId/words', 'admin/:categoryId/words'])
+  async allWords(@Param('categoryId') categoryId: number) {
+    return await this.categoriesService.findOne({
+      where: {
+        category_id: categoryId,
+      },
       relations: ['words'],
     });
   }
-
+  // Reminder: Add authguard
   @Post('admin/category/create')
   async create(@Body() body: CreateCategoryDto) {
     return this.categoriesService.save(body);
   }
+
+  // Reminder: Add authguard
+  @Get([
+    'student/:categoryId/words/:wordsId',
+    'admin/:categoryId/words/:wordsId',
+  ])
+  async getChoices(
+    @Param('categoryId') categoryId: number,
+    @Param('wordsId') wordsId: number,
+  ) {
+    const words = await this.wordsService.findOne({
+      where: { word_id: wordsId, category_id: categoryId },
+      relations: ['choices'],
+    });
+
+    if (words && categoryId) {
+      const formattedData = {
+        word_id: words.word_id,
+        category_id: words.category_id,
+        given_word: words.given_word,
+        correct_word: words.correct_word,
+        choices: words.choices.map((choice) => {
+          return {
+            id: choice.id,
+            options: choice.options,
+          };
+        }),
+      };
+
+      return formattedData;
+    }
+
+    throw new NotFoundException('Category or words not found!');
+  }
+  // Reminder: Add authguard
+  @Post('admin/category/:category_id/add')
+  async createWord(
+    @Param('category_id') category_id: number,
+    @Body() body: AddWordDto,
+  ) {
+    const category = await this.categoriesService.findOne({
+      where: { category_id },
+    });
+
+    if (!category) {
+      throw new NotFoundException('Category not found!');
+    }
+    return this.wordsService.save({
+      category_id,
+      given_word: body.given_word,
+      correct_word: body.correct_word,
+      choices: body.choices,
+    });
+  }
+
+  @Put('admin/category/:category_id/edit')
+  async editCategory(
+    @Param('category_id') category_id: number,
+    @Body('title') title: string,
+    @Body('description') description: string,
+  ) {
+    const category = await this.categoriesService.findOne({
+      where: { category_id },
+    });
+
+    if (!category) {
+      throw new NotFoundException('Category not found!');
+    }
+
+    await this.categoriesService.update(category_id, {
```

Comment:
Does `update()` return a `category`?

If so, I think you can just return this instead of finding again in **line: 120**

Reply:
@RDLegs  removing the line 120 sir will result to this 

![image](https://github.com/framgia/sph-els-mark/assets/110364391/ec93b324-cb40-4adc-b622-805b354de7c1)

Reply:
I guess it's fine to find again for now.